File: 2017/day22.js
/*

--- Day 22: Sporifica Virus ---

Diagnostics indicate that the local grid computing cluster has been contaminated with the Sporifica Virus. The grid computing cluster is a seemingly-infinite two-dimensional grid of compute nodes. Each node is either clean or infected by the virus.

To prevent overloading the nodes (which would render them useless to the virus) or detection by system administrators, exactly one virus carrier moves through the network, infecting or cleaning nodes as it moves. The virus carrier is always located on a single node in the network (the current node) and keeps track of the direction it is facing.

To avoid detection, the virus carrier works in bursts; in each burst, it wakes up, does some work, and goes back to sleep. The following steps are all executed in order one time each burst:

If the current node is infected, it turns to its right. Otherwise, it turns to its left. (Turning is done in-place; the current node does not change.)
If the current node is clean, it becomes infected. Otherwise, it becomes cleaned. (This is done after the node is considered for the purposes of changing direction.)
The virus carrier moves forward one node in the direction it is facing.
Diagnostics have also provided a map of the node infection status (your puzzle input). Clean nodes are shown as .; infected nodes are shown as #. This map only shows the center of the grid; there are many more nodes beyond those shown, but none of them are currently infected.

The virus carrier begins in the middle of the map facing up.

For example, suppose you are given a map like this:

..#
#..
...

Then, the middle of the infinite grid looks like this, with the virus carrier's position marked with [ ]:

. . . . . . . . .
. . . . . . . . .
. . . . . . . . .
. . . . . # . . .
. . . #[.]. . . .
. . . . . . . . .
. . . . . . . . .
. . . . . . . . .

The virus carrier is on a clean node, so it turns left, infects the node, and moves left:

. . . . . . . . .
. . . . . . . . .
. . . . . . . . .
. . . . . # . . .
. . .[#]# . . . .
. . . . . . . . .
. . . . . . . . .
. . . . . . . . .

The virus carrier is on an infected node, so it turns right, cleans the node, and moves up:

. . . . . . . . .
. . . . . . . . .
. . . . . . . . .
. . .[.]. # . . .
. . . . # . . . .
. . . . . . . . .
. . . . . . . . .
. . . . . . . . .

Four times in a row, the virus carrier finds a clean, infects it, turns left, and moves forward, ending in the same place and still facing up:

. . . . . . . . .
. . . . . . . . .
. . . . . . . . .
. . #[#]. # . . .
. . # # # . . . .
. . . . . . . . .
. . . . . . . . .
. . . . . . . . .

Now on the same node as before, it sees an infection, which causes it to turn right, clean the node, and move forward:

. . . . . . . . .
. . . . . . . . .
. . . . . . . . .
. . # .[.]# . . .
. . # # # . . . .
. . . . . . . . .
. . . . . . . . .
. . . . . . . . .

After the above actions, a total of 7 bursts of activity had taken place. Of them, 5 bursts of activity caused an infection.

After a total of 70, the grid looks like this, with the virus carrier facing up:

. . . . . # # . .
. . . . # . . # .
. . . # . . . . #
. . # . #[.]. . #
. . # . # . . # .
. . . . . # # . .
. . . . . . . . .
. . . . . . . . .

By this time, 41 bursts of activity caused an infection (though most of those nodes have since been cleaned).

After a total of 10000 bursts of activity, 5587 bursts will have caused an infection.

Given your actual map, after 10000 bursts of activity, how many bursts cause a node to become infected? (Do not count nodes that begin infected.)


--- Part Two ---

As you go to remove the virus from the infected nodes, it evolves to resist your attempt.

Now, before it infects a clean node, it will weaken it to disable your defenses. If it encounters an infected node, it will instead flag the node to be cleaned in the future. So:

Clean nodes become weakened.
Weakened nodes become infected.
Infected nodes become flagged.
Flagged nodes become clean.

Every node is always in exactly one of the above states.

The virus carrier still functions in a similar way, but now uses the following logic during its bursts of action:

Decide which way to turn based on the current node:
If it is clean, it turns left.
If it is weakened, it does not turn, and will continue moving in the same direction.
If it is infected, it turns right.
If it is flagged, it reverses direction, and will go back the way it came.
Modify the state of the current node, as described above.
The virus carrier moves forward one node in the direction it is facing.

Start with the same map (still using . for clean and # for infected) and still with the virus carrier starting in the middle and facing up.

Using the same initial state as the previous example, and drawing weakened as W and flagged as F, the middle of the infinite grid looks like this, with the virus carrier's position again marked with [ ]:

. . . . . . . . .
. . . . . . . . .
. . . . . . . . .
. . . . . # . . .
. . . #[.]. . . .
. . . . . . . . .
. . . . . . . . .
. . . . . . . . .

This is the same as before, since no initial nodes are weakened or flagged. The virus carrier is on a clean node, so it still turns left, instead weakens the node, and moves left:

. . . . . . . . .
. . . . . . . . .
. . . . . . . . .
. . . . . # . . .
. . .[#]W . . . .
. . . . . . . . .
. . . . . . . . .
. . . . . . . . .

The virus carrier is on an infected node, so it still turns right, instead flags the node, and moves up:

. . . . . . . . .
. . . . . . . . .
. . . . . . . . .
. . .[.]. # . . .
. . . F W . . . .
. . . . . . . . .
. . . . . . . . .
. . . . . . . . .

This process repeats three more times, ending on the previously-flagged node and facing right:

. . . . . . . . .
. . . . . . . . .
. . . . . . . . .
. . W W . # . . .
. . W[F]W . . . .
. . . . . . . . .
. . . . . . . . .
. . . . . . . . .

Finding a flagged node, it reverses direction and cleans the node:

. . . . . . . . .
. . . . . . . . .
. . . . . . . . .
. . W W . # . . .
. .[W]. W . . . .
. . . . . . . . .
. . . . . . . . .
. . . . . . . . .

The weakened node becomes infected, and it continues in the same direction:

. . . . . . . . .
. . . . . . . . .
. . . . . . . . .
. . W W . # . . .
.[.]# . W . . . .
. . . . . . . . .
. . . . . . . . .
. . . . . . . . .

Of the first 100 bursts, 26 will result in infection. Unfortunately, another feature of this evolved virus is speed; of the first 10000000 bursts, 2511944 will result in infection.

Given your actual map, after 10000000 bursts of activity, how many bursts cause a node to become infected? (Do not count nodes that begin infected.)

*/

// OPTIONAL VARIABLES
const DISPLAY_EXTRA_INFO = true;

function traverseInfiniteGridWhereCoordsHaveMultipleStates (part, inputStr, extraParam, DEBUG = false) {
  const inputArr = inputStr.split('\r\n');
  
  const H = inputArr.length;
  const W = inputArr[0].length;

  const INFECTED = '#';
  const INFECTED_SET = new Set();
  const WEAKENED_SET = new Set();                                                 // for part 2
  const FLAGGED_SET = new Set();                                                  // for part 2

  const NUM_ROUNDS = extraParam;
  const DELTAS = [
    [ -1, 0 ],
    [ 0, +1 ],
    [ +1, 0 ],
    [ 0, -1 ],
  ];

  // PARSE INPUT DATA
  for (let row = 0; row < H; ++row) {
    for (let col = 0; col < W; ++col) {
      if (inputArr[row][col] === INFECTED) {
        const coords = `${row - Math.floor(H/2)},${col - Math.floor(W/2)}`;       // add #s to infected set
        INFECTED_SET.add(coords);
      }
    }
  }


  // INIT
  let [ y, x ] = [ 0, 0 ];
  let dir = 0;
  let roundsCausingInfection = 0;

  // ANALYZE
  const TIME_AT_START = Date.now();
  console.log('RUNNING ANALYSIS (PLEASE WAIT)...');
  for (let i = 0; i < NUM_ROUNDS; ++i) {
    const coords = `${y},${x}`;

    if (part === 1) {                                                             // PART 1: ONLY CLEAN AND INFECTED STATES

      if (INFECTED_SET.has(coords)) {
        dir = (dir + 1) % 4;                                                      // turn right
        INFECTED_SET.delete(coords);                                              // infected --> clean
      } else {
        dir = (dir + 4 - 1) % 4;                                                  // turn left
        INFECTED_SET.add(coords);                                                 // clean --> infected
        ++roundsCausingInfection;                                                 // infection is caused from a clean state
      }

    } else {                                                                      // PART 2: ALSO WEAKENED AND FLAGGED STATES

      if (INFECTED_SET.has(coords)) {
        dir = (dir + 1) % 4;                                                      // turn right
        INFECTED_SET.delete(coords);
        FLAGGED_SET.add(coords);                                                  // infected --> flagged
      } else if (FLAGGED_SET.has(coords)) {
        dir = (dir + 2) % 4;                                                      // turn around
        FLAGGED_SET.delete(coords);                                               // flagged --> clean
      } else if (WEAKENED_SET.has(coords)) {
        WEAKENED_SET.delete(coords);
        INFECTED_SET.add(coords);                                                 // weakened --> infected
        ++roundsCausingInfection;                                                 // infection is caused from a weakened state
      } else {
        dir = (dir + 4 - 1) % 4;                                                  // turn left
        WEAKENED_SET.add(coords);                                                 // clean --> weakened
      }

    }

    const [ dy, dx ] = DELTAS[dir];
    y += dy;
    x += dx;
  }

  console.log(`(RUN TOOK ${(Date.now() - TIME_AT_START)/1000} SECS)`);
  return roundsCausingInfection;
}

// TEST CASES

const test = require('./_test');
const testNum = [1];
let input, expected;
const func = traverseInfiniteGridWhereCoordsHaveMultipleStates;
const sortedFunc = (...args) => func(...args).sort();                   // used when the order of the output does not matter
const modFunc = (...args) => func(...args) % 1000000007;                // used when the output is very large
const skippedTests = new Set([  ]);
const lowestTest = 0;
const highestTest = 0;

const fs = require('fs');
const path = require('path');
const DAY_NUM = __filename.split('.js')[0].split('day')[1];
const INPUT_PATH = path.join(__dirname, `day${DAY_NUM}-input.txt`);
const actualInput = fs.readFileSync(INPUT_PATH, 'utf8');
const parseSampleInput = s => s.split('').map(c => c === '\n' ? '\r\n' : c).join('');

const sampleInput = parseSampleInput(
`..#
#..
...`
);

// Test case 1
input = {
  part: 1,
  inputStr: sampleInput,
  extraParam: 7,
  DEBUG: true,
};
expected = 5;
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

// Test case 2
input = {
  part: 1,
  inputStr: sampleInput,
  extraParam: 70,
  DEBUG: true,
};
expected = 41;
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

// Test case 3
input = {
  part: 1,
  inputStr: sampleInput,
  extraParam: 10000,
  DEBUG: true,
};
expected = 5587;
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

// Test case 4
input = {
  part: 1,
  inputStr: actualInput,
  extraParam: 10000,
};
expected = 5261;
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

// Test case 5
input = {
  part: 2,
  inputStr: sampleInput,
  extraParam: 100,
  DEBUG: true,
};
expected = 26;
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

// Test case 6
input = {
  part: 2,
  inputStr: sampleInput,
  extraParam: 10000000,
  DEBUG: true,
};
expected = 2511944;
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

// Test case 7
input = {
  part: 2,
  inputStr: actualInput,
  extraParam: 10000000,
};
expected = 2511927;
test(func, input, expected, testNum, skippedTests, lowestTest, highestTest);

File: 2023/day22.py
"""

--- Day 22: Sand Slabs ---

Enough sand has fallen; it can finally filter water for Snow Island.

Well, almost.

The sand has been falling as large compacted bricks of sand, piling up to form an impressive stack here near the edge of Island Island. In order to make use of the sand to filter water, some of the bricks will need to be broken apart - nay, disintegrated - back into freely flowing sand.

The stack is tall enough that you'll have to be careful about choosing which bricks to disintegrate; if you disintegrate the wrong brick, large portions of the stack could topple, which sounds pretty dangerous.

The Elves responsible for water filtering operations took a snapshot of the bricks while they were still falling (your puzzle input) which should let you work out which bricks are safe to disintegrate. For example:

1,0,1~1,2,1
0,0,2~2,0,2
0,2,3~2,2,3
0,0,4~0,2,4
2,0,5~2,2,5
0,1,6~2,1,6
1,1,8~1,1,9

Each line of text in the snapshot represents the position of a single brick at the time the snapshot was taken. The position is given as two x,y,z coordinates - one for each end of the brick - separated by a tilde (~). Each brick is made up of a single straight line of cubes, and the Elves were even careful to choose a time for the snapshot that had all of the free-falling bricks at integer positions above the ground, so the whole snapshot is aligned to a three-dimensional cube grid.

A line like 2,2,2~2,2,2 means that both ends of the brick are at the same coordinate - in other words, that the brick is a single cube.

Lines like 0,0,10~1,0,10 or 0,0,10~0,1,10 both represent bricks that are two cubes in volume, both oriented horizontally. The first brick extends in the x direction, while the second brick extends in the y direction.

A line like 0,0,1~0,0,10 represents a ten-cube brick which is oriented vertically. One end of the brick is the cube located at 0,0,1, while the other end of the brick is located directly above it at 0,0,10.

The ground is at z=0 and is perfectly flat; the lowest z value a brick can have is therefore 1. So, 5,5,1~5,6,1 and 0,2,1~0,2,5 are both resting on the ground, but 3,3,2~3,3,3 was above the ground at the time of the snapshot.

Because the snapshot was taken while the bricks were still falling, some bricks will still be in the air; you'll need to start by figuring out where they will end up. Bricks are magically stabilized, so they never rotate, even in weird situations like where a long horizontal brick is only supported on one end. Two bricks cannot occupy the same position, so a falling brick will come to rest upon the first other brick it encounters.

Here is the same example again, this time with each brick given a letter so it can be marked in diagrams:

1,0,1~1,2,1   <- A
0,0,2~2,0,2   <- B
0,2,3~2,2,3   <- C
0,0,4~0,2,4   <- D
2,0,5~2,2,5   <- E
0,1,6~2,1,6   <- F
1,1,8~1,1,9   <- G

At the time of the snapshot, from the side so the x axis goes left to right, these bricks are arranged like this:

 x
012
.G. 9
.G. 8
... 7
FFF 6
..E 5 z
D.. 4
CCC 3
BBB 2
.A. 1
--- 0

Rotating the perspective 90 degrees so the y axis now goes left to right, the same bricks are arranged like this:

 y
012
.G. 9
.G. 8
... 7
.F. 6
EEE 5 z
DDD 4
..C 3
B.. 2
AAA 1
--- 0

Once all of the bricks fall downward as far as they can go, the stack looks like this, where ? means bricks are hidden behind other bricks at that location:

 x
012
.G. 6
.G. 5
FFF 4
D.E 3 z
??? 2
.A. 1
--- 0

Again from the side:

 y
012
.G. 6
.G. 5
.F. 4
??? 3 z
B.C 2
AAA 1
--- 0

Now that all of the bricks have settled, it becomes easier to tell which bricks are supporting which other bricks:

    Brick A is the only brick supporting bricks B and C.
    Brick B is one of two bricks supporting brick D and brick E.
    Brick C is the other brick supporting brick D and brick E.
    Brick D supports brick F.
    Brick E also supports brick F.
    Brick F supports brick G.
    Brick G isn't supporting any bricks.

Your first task is to figure out which bricks are safe to disintegrate. A brick can be safely disintegrated if, after removing it, no other bricks would fall further directly downward. Don't actually disintegrate any bricks - just determine what would happen if, for each brick, only that brick were disintegrated. Bricks can be disintegrated even if they're completely surrounded by other bricks; you can squeeze between bricks if you need to.

In this example, the bricks can be disintegrated as follows:

    Brick A cannot be disintegrated safely; if it were disintegrated, bricks B and C would both fall.
    Brick B can be disintegrated; the bricks above it (D and E) would still be supported by brick C.
    Brick C can be disintegrated; the bricks above it (D and E) would still be supported by brick B.
    Brick D can be disintegrated; the brick above it (F) would still be supported by brick E.
    Brick E can be disintegrated; the brick above it (F) would still be supported by brick D.
    Brick F cannot be disintegrated; the brick above it (G) would fall.
    Brick G can be disintegrated; it does not support any other bricks.

So, in this example, 5 bricks can be safely disintegrated.

Figure how the blocks will settle based on the snapshot. Once they've settled, consider disintegrating a single brick; how many bricks could be safely chosen as the one to get disintegrated?


--- Part Two ---

Disintegrating bricks one at a time isn't going to be fast enough. While it might sound dangerous, what you really need is a chain reaction.

You'll need to figure out the best brick to disintegrate. For each brick, determine how many other bricks would fall if that brick were disintegrated.

Using the same example as above:

    Disintegrating brick A would cause all 6 other bricks to fall.
    Disintegrating brick F would cause only 1 other brick, G, to fall.

Disintegrating any other brick would cause no other bricks to fall. So, in this example, the sum of the number of other bricks that would fall as a result of disintegrating each brick is 7.

For each brick, determine how many other bricks would fall if that brick were disintegrated. What is the sum of the number of other bricks that would fall?


"""

# LIBRARIES
import time
import pathlib
import os
from math import *
from functools import *
from sys import *
from collections import *
from queue import *

# MODULES
from _test import test

# OPTIONAL VARIABLES
DISPLAY_EXTRA_INFO = True
# DISPLAY_EXTRA_INFO = False

def analyze_falling_blocks(part, input_str, DEBUG = False):

  TIME_AT_START = time.time()

  # CONSTANTS

  X, Y, Z = 'x', 'y', 'z'

  # DATA STRUCTURES

  MAP = set()                                                       # holds tuples of coordinates that are occupied by some block  
  BRICKS = []                                                       # holds collections of bricks with their data


  # PARSE INPUT DATA

  input_arr = input_str.split('\n')
  for line in input_arr:
    left_str, right_str = line.split('~')
    [ xl, yl, zl ] = [ int(n) for n in left_str.split(',') ]
    [ xr, yr, zr ] = [ int(n) for n in right_str.split(',') ]
    
    dx, dy, dz = abs(xl - xr), abs(yl - yr), abs(zl - zr)

    brick = {
      'points': [],
      'lowest': zl if dz == 0 else min(zl, zr),                     # helps to sort bricks in order of their lowest block when dropping all bricks
      'orientation': None                                           # helps to know if a brick is in z orientation, so its droppability is determined
                                                                    # solely by whether space is clear below its lowest point only
    }

    if dx + dy + dz == 0:                                           # EDGE CASE: brick is 1x1x1 (because dx, dy, and dz are all 0)
      brick['points'].append([ xl, yl, zl ])

    elif dx != 0:                                                   # x orientation
      brick['orientation'] = X
      for x in range(min(xl, xr), max(xl, xr) + 1):                 # save each block into brick data, and occupy the MAP
        brick['points'].append([x, yl, zl])
        MAP.add((x, yl, zl))
    
    elif dy != 0:                                                   # y orientation
      brick['orientation'] = Y
      for y in range(min(yl, yr), max(yl, yr) + 1):
        brick['points'].append([xl, y, zl])
        MAP.add((xl, y, zl))
    
    elif dz != 0:                                                   # z orientation
      brick['orientation'] = Z
      for z in range(min(zl, zr), max(zl, zr) + 1):
        brick['points'].append([xl, yl, z])
        MAP.add((xl, yl, z))
      brick['points'].sort(key=lambda point: point[2])              # sort the points such that brick['points'][0] is the lowest point
    
    else: assert False

    BRICKS.append(brick)

  BRICKS.sort(key=lambda brick: brick['lowest'])                    # sort the bricks such that BRICKS[0]'s lowest point is the lowest of them all


  # HELPER FUNCTIONS

  def drop_bricks(BRICKS, MAP):                                     # drops all bricks into a static position, and returns the number of bricks that moved in the process

    bricks_that_fell = 0

    for brick in BRICKS:
      brick_fell = False                                            # track whether this brick fell
      while True:                                                   # keep iterating while this brick can fall
        can_fall = True

        if brick['points'][0][2] == 1:                              # CASE 1: brick's lowest point is already resting on the floor
                                                                    # (this works because: if Z orientation, we sorted blocks earlier; else, all blocks have same Z)
          can_fall = False

        elif brick['orientation'] == Z:                             # CASE 2: Z oriented brick: ONLY check space below bottom-most point
          bottom_brick = brick['points'][0]                         # this works because we sorted it during preprocessing!
          [x, y, z] = bottom_brick
          if (x, y, z - 1) in MAP:                                  # negate can_fall if the space below the bottom point is occupied
            can_fall = False
        
        else:                                                       # CASE 3: check every point along the block to see if any of them is resting on an occupied point
          for [x, y, z] in brick['points']:
            if (x, y, z - 1) in MAP:
              can_fall = False
              break

        if can_fall == False:                                       # if this brick cannot fall, break out of the while loop
          break
        
        brick_fell = True                                           # if we did not immediately break out of the while loop, this brick fell at least one unit distance
        lowest = float('inf')                                       # OPTIONAL: update information about its lowest point (optional because we no longer use it after preprocessing)
        for i in range(len(brick['points'])):
          [x, y, z] = brick['points'][i]
          brick['points'][i] = [x, y, z - 1]
          MAP.add((x, y, z - 1))
          # MAP.remove((x, y, z))
          MAP.discard((x, y, z))
          lowest = min(lowest, z - 1)                               # OPTIONAL: update information about its lowest point (optional because we no longer use it after preprocessing)

        brick['lowest'] = lowest                                    # OPTIONAL: update information about its lowest point (optional because we no longer use it after preprocessing)

      if brick_fell: bricks_that_fell += 1

    return bricks_that_fell
  

  def deep_copy_bricks(BRICKS):
    copy = []
    for brick in BRICKS:
      brick_copy = { 'points': [ point.copy() for point in brick['points'] ], 'lowest': brick['lowest'], 'orientation': brick['orientation'] }
      copy.append(brick_copy)
    return copy


  # PRE-PROCESSING: DROP ALL BRICKS INTO A STATIC POSITION

  drop_bricks(BRICKS, MAP)


  if not DEBUG: print(f"(TOOK {(time.time() - TIME_AT_START)} SECS TO DO ALL PREPROCESSING)")


  # ANALYZE

  TIME_AT_START_OF_ANALYSIS = time.time()

  if part == 1:                                                                             # PART 1: COUNT THE NUMBER OF BRICKS THAT CAN BE DISINTEGRATED WITH NOTHING ELSE FALLING

    if not DEBUG: print('RUNNING PART 1 ANALYSIS (PLEASE WAIT)...')

    num_can_safely_disintegrate = 0

    for i in range(len(BRICKS)):

      # IMPORTANT: MAKE DEEP COPY TO GET SPLICE OF BRICKS WITHOUT CURRENT BRICK
      deep_copy = deep_copy_bricks(BRICKS)
      bricks_copy = deep_copy[:i] + deep_copy[i+1:]

      # IMPORTANT: MAKE DEEP COPY OF MAP
      MAP_copy = MAP.copy()
      brick_removed = deep_copy[i]
      for (x, y, z) in brick_removed['points']:
        MAP_copy.remove((x, y, z))

      # Use helper function with deep copied data structures
      if drop_bricks(bricks_copy, MAP_copy) == 0:
        num_can_safely_disintegrate += 1

    if not DEBUG: print(f"(RUN TOOK {(time.time() - TIME_AT_START_OF_ANALYSIS)} SECS)")
    return num_can_safely_disintegrate

  else:                                                                                     # PART 2: SUM THE NUMBER OF OTHER BRICKS THAT WOULD FALL FOR EACH BRICK THAT GETS DISINTEGRATED

    if not DEBUG: print('RUNNING PART 2 ANALYSIS (PLEASE WAIT)...')

    total = 0

    for i in range(len(BRICKS)):

      # IMPORTANT: MAKE DEEP COPY TO GET SPLICE OF BRICKS WITHOUT CURRENT BRICK
      deep_copy = deep_copy_bricks(BRICKS)
      bricks_copy = deep_copy[:i] + deep_copy[i+1:]

      # IMPORTANT: MAKE DEEP COPY OF MAP
      MAP_copy = MAP.copy()
      brick_removed = deep_copy[i]
      for (x, y, z) in brick_removed['points']:
        MAP_copy.remove((x, y, z))

      # Use helper function with deep copied data structures
      total += drop_bricks(bricks_copy, MAP_copy)

    if not DEBUG: print(f"(RUN TOOK {(time.time() - TIME_AT_START_OF_ANALYSIS)} SECS)")
    return total


# TEST CASES

test_num = [1]
test_input = None
test_expected = None
func = analyze_falling_blocks
skipped_tests = set([ 2, 3, 4 ])
skipped_tests = set([ 3, 4 ])
skipped_tests = set([ 4 ])
skipped_tests = set([  ])
lowest_test = 0
highest_test = 0

FILENAME = pathlib.Path(__file__).name
FILENAME_SPLIT = FILENAME.split('.py')[0].split('day')
DAY_NUM = FILENAME_SPLIT[1] if len(FILENAME_SPLIT) > 1 else None
CURRENT_DIR = os.path.dirname(os.path.realpath(__file__))
INPUT_PATH = CURRENT_DIR + '\\' + (f"day{DAY_NUM}-input.txt" if DAY_NUM else 'template-input.txt')
actual_input = open(INPUT_PATH, 'r').read()

sample_input = """1,0,1~1,2,1
0,0,2~2,0,2
0,2,3~2,2,3
0,0,4~0,2,4
2,0,5~2,2,5
0,1,6~2,1,6
1,1,8~1,1,9"""

# Test case 1
test_input = {
  'part': 1,
  'input_str': sample_input,
  'DEBUG': True,
}
test_expected = 5
test(func, test_input, test_expected, test_num, skipped_tests, lowest_test, highest_test)

# Test case 2
test_input = {
  'part': 1,
  'input_str': actual_input,
}
test_expected = 475
test(func, test_input, test_expected, test_num, skipped_tests, lowest_test, highest_test)

# Test case 3
test_input = {
  'part': 2,
  'input_str': sample_input,
  'DEBUG': True,
}
test_expected = 7
test(func, test_input, test_expected, test_num, skipped_tests, lowest_test, highest_test)

# Test case 4
test_input = {
  'part': 2,
  'input_str': actual_input,
}
test_expected = 79144
test(func, test_input, test_expected, test_num, skipped_tests, lowest_test, highest_test)